Use spawnSync instead of execaSync in `git.ts`

kevinzunigacuellar opened the pull request:
<!-- Thank you for opening a PR! We really appreciate you taking the time to help out 🙌 -->

#### Description

- Closes #887
- Replace `execa` with `node:child_process`

<!--
Here’s what will happen next:
One or more of our maintainers will take a look and may ask you to make changes.
We try to be responsive, but don’t worry if this takes a day or two.
-->

astrobot-houston commented:
## size-limit report 📦 
| Path          | Size          |
| ------------- | ------------- |
| /index.html   | 5.21 KB (0%)  |
| /_astro/*.js  | 21.42 KB (0%) |
| /_astro/*.css | 13.04 KB (0%) |

HiDeoo commented:
This is super cool, I remember starting something similar a while back but never finished it. Glad to see someone tackling this :raised_hands:

A few notes:

- I'm not sure this should spawn a new shell so I think `spawnSync` (which accepts a `encoding: 'utf-8'` option to get a string back in `result.stdout`) might be a better choice and would also not change the current behavior.
- `execa` should be removed.
- Maybe we could take this opportunity to remove the `oldest` code path as we do not ever use it (and maybe rename the function? I remember using  `getFileNewestCommitDate` at the time) and also remove `FileNotTrackedError` and replace it by a basic `Error`?
- I remember starting to work on a `git.test.ts` file for this as we started to own a bit more this piece of code. If you think this could be useful, I'm sharing the WIP here in case it could help (this assume `git` is installed on the machine running the tests but I think this is a fair assumption?)

  ```ts
  import { mkdtempSync, writeFileSync } from 'node:fs';
  import { join } from 'node:path';
  import { tmpdir } from 'node:os';
  import { spawnSync } from 'node:child_process';
  import { describe, expect, test } from 'vitest';
  import { getFileNewestCommitDate } from '../../utils/git';

  describe('getFileNewestCommitDate', () => {
    const { commitAllChanges, getFilePath, writeFile } = makeTestRepo();

    test('returns the newest commit date', () => {
      const file = 'updated.md';
      const lastCommitDate = '2023-06-25';

      writeFile(file, 'content 0');
      commitAllChanges('add updated.md', '2023-06-21');
      writeFile(file, 'content 1');
      commitAllChanges('update updated.md', lastCommitDate);

      expectCommitDateToEqual(getFileNewestCommitDate(getFilePath(file)), lastCommitDate);
    });

    test('returns the initial commit date for a file never updated', () => {
      const file = 'added.md';
      const commitDate = '2022-09-18';

      writeFile(file, 'content');
      commitAllChanges('add added.md', commitDate);

      expectCommitDateToEqual(getFileNewestCommitDate(getFilePath(file)), commitDate);
    });

    test('returns the newest commit date for a file with a name that contains a space', () => {
      const file = 'updated with space.md';
      const lastCommitDate = '2021-01-02';

      writeFile(file, 'content 0');
      commitAllChanges('add updated.md', '2021-01-01');
      writeFile(file, 'content 1');
      commitAllChanges('update updated.md', lastCommitDate);

      expectCommitDateToEqual(getFileNewestCommitDate(getFilePath(file)), lastCommitDate);
    });

    test('returns the newest commit date for a file updated the same day', () => {
      const file = 'updated-same-day.md';
      const lastCommitDate = '2023-06-25T14:22:35Z';

      writeFile(file, 'content 0');
      commitAllChanges('add updated.md', '2023-06-25T12:34:56Z');
      writeFile(file, 'content 1');
      commitAllChanges('update updated.md', lastCommitDate);

      expectCommitDateToEqual(getFileNewestCommitDate(getFilePath(file)), lastCommitDate);
    });

    test('throws when failing to retrieve the git history for a file', () => {
      expect(() =>
        getFileNewestCommitDate(getFilePath('../not-a-starlight-test-repo/test.md'))
      ).toThrow(/^Failed to retrieve the git history for file "[/\\-\w ]+\/test\.md" with exit code/);
    });

    test('throws when trying to get the history of a non-existing or untracked file', () => {
      const expectedError =
        /^Failed to retrieve the git history for file "[/\\-\w ]+\/(?:unknown|untracked)\.md" because the file is not tracked by git\.$/;

      writeFile('untracked.md', 'content');

      expect(() => getFileNewestCommitDate(getFilePath('unknown.md'))).toThrow(expectedError);
      expect(() => getFileNewestCommitDate(getFilePath('untracked.md'))).toThrow(expectedError);
    });
  });

  function expectCommitDateToEqual(commitDate: CommitDate, expectedDateStr: ISODate) {
    const expectedDate = new Date(expectedDateStr);

    expect(commitDate).toStrictEqual({
      date: expectedDate,
      timestamp: expectedDate.getTime() / 1000,
    });
  }

  function makeTestRepo() {
    const repoPath = mkdtempSync(join(tmpdir(), 'starlight-test-git-'));

    function runInRepo(command: string, args: string[], env: NodeJS.ProcessEnv = {}) {
      const result = spawnSync(command, args, { cwd: repoPath, env });

      if (result.status !== 0) {
        throw new Error(`Failed to execute test repository command: '${command} ${args.join(' ')}'`);
      }
    }

    // Configure git specifically for this test repository.
    runInRepo('git', ['init']);
    runInRepo('git', ['config', 'user.name', 'starlight-test']);
    runInRepo('git', ['config', 'user.email', 'starlight-test@example.com']);
    runInRepo('git', ['config', 'commit.gpgsign', 'false']);

    return {
      // The `dateStr` argument should be in the `YYYY-MM-DD` or `YYYY-MM-DDTHH:MM:SSZ` format.
      commitAllChanges(message: string, dateStr: ISODate) {
        const date = dateStr.endsWith('Z') ? dateStr : `${dateStr}T00:00:00Z`;

        runInRepo('git', ['add', '-A']);
        // This sets both the author and committer dates to the provided date.
        runInRepo('git', ['commit', '-m', message, '--date', date], { GIT_COMMITTER_DATE: date });
      },
      getFilePath(name: string) {
        return join(repoPath, name);
      },
      writeFile(name: string, content: string) {
        writeFileSync(join(repoPath, name), content);
      },
    };
  }

  type ISODate =
    | `${number}-${number}-${number}`
    | `${number}-${number}-${number}T${number}:${number}:${number}Z`;

  type CommitDate = ReturnType<typeof getFileNewestCommitDate>;
  ```

kevinzunigacuellar commented:
Thanks for the review Hideo 💜 . I added most of your suggested changes, and cleaned up the logic a bit. However, I seem to have broken a test 😆, which I am not sure how to fix. It seems that my root file is pointing to the test directory but every time I try changing root path in the `test-config` I break the entire thing 💥. Do you have any ideas?

HiDeoo commented:
I don't have a computer to actually run the code but from the changes, I would say that reverting the changes to `getLastUpdated()` (minus the name and argument changes to `getFileCommitDate()` ofc) should fix the issue as you removed the try-catch block that we would still need to ignore the error and I think that in this test, the path is invalid as it's based on the actual docs (that's also one of the reason I started working on the dedicated tests I linked earlier so we can easily test the git part in isolation).

kevinzunigacuellar commented:
Thank you Hideo, it was the error catch 😄 . I was able to add your tests in `git.test.ts` 💜

delucis commented on packages/starlight/utils/route-data.ts:
```diff
@@ -70,17 +70,24 @@ function getToC({ entry, locale, headings }: PageProps) {
-			try {
-				({ date } = getFileCommitDate(currentFilePath, 'newest'));
-			} catch {}
+		let date: Date | undefined;
+		try {
+			date =
+				frontmatterLastUpdated instanceof Date
+					? frontmatterLastUpdated
+					: getNewestCommitDate(currentFilePath);
+		} catch {
+			// If the git command fails, ignore the error and continue.
 		}
 		return date;
```
If we’re doing it this way, I guess we can do:
```suggestion
		try {
			return frontmatterLastUpdated instanceof Date
					? frontmatterLastUpdated
					: getNewestCommitDate(currentFilePath);
		} catch {
			// If the git command fails, ignore the error.
			return undefined;
		}
```

delucis reviewed (commented):
Woo this looks great. Thanks @kevinzunigacuellar and thank you @HiDeoo for your very insightful review 💖 

Left one small refactor suggestion, but overall, the code looks good to me.

One thought: have we tested this on Windows? We don’t currently use the last-updated feature on the docs site and building the Starlight docs is the only “test” we currently run on Windows. Would be good to be sure that’s working.

lilnasy reviewed (approved):
Haven't looked at the code but the result looks great to me!
<details>
<summary>steps chris told me to follow</summary>
<img width="497" alt="image" src="https://github.com/withastro/starlight/assets/69170106/51367091-3801-4964-ae5c-1caaafbfdd43">
</details>

<details>
<summary>result</summary>
<img width="564" alt="image" src="https://github.com/withastro/starlight/assets/69170106/58a77637-133d-45a7-b64c-004f5b13abf1">

</details>

hippotastic commented:
Also works well for me on Windows 10!

at-the-vr commented:
Working for me as well on Win 10, I went across the translations found no Issues. Found the locale translation interesting and that's without any hardcoding 🔥🔥
| Locale  | Last Updated Image |
| ------------- | ------------- |
| Deutsch  | ![image](https://github.com/withastro/starlight/assets/88548999/17b16671-b54f-4475-83b5-a0c0faa11633)  |
| Japanese & Korean  | ![image](https://github.com/withastro/starlight/assets/88548999/709eaca5-8019-47d4-bcc2-010ef670f138)|

There are 13 more languages but these 3 caught my attention 😅. Overall it works 😄

delucis reviewed (approved):
Let’s go! Thanks again @kevinzunigacuellar and thanks everyone for testing this out 💖
